File: apps/api/app/engine.py
"""Deterministic demo engine — mirrors data/demo/generate_seed.py exactly.

This is NOT a trained model. Scores are simulated and labelled as such.
Thresholds come from data/demo/policy.json and are prototype policies only.
"""
from __future__ import annotations

import math
from typing import Any


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    r = 6371.0
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dp = math.radians(lat2 - lat1)
    dl = math.radians(lon2 - lon1)
    a = math.sin(dp / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2) ** 2
    return 2 * r * math.asin(math.sqrt(a))


class DeterministicEngine:
    def __init__(self, policy: dict[str, Any], taxonomy: dict[str, Any]):
        self.policy = policy
        self.taxonomy = taxonomy
        self.conditions = {c["id"]: c for c in taxonomy["conditions"]}

    # ---------------- capture quality gate ----------------
    def capture_quality(self, checklist: dict[str, bool]) -> dict[str, Any]:
        cq = self.policy["captureQuality"]
        w = cq["weights"]
        cov = round(sum(w[k] for k in ("leafClose", "lowerLeaf", "wholePlant", "lightingOk") if checklist.get(k)), 2)
        issues: list[str] = []
        reqs: list[str] = []
        if not checklist.get("leafClose"):
            issues.append("Missing close-up of affected leaf")
            reqs.append("leafClose")
        if not checklist.get("lowerLeaf"):
            issues.append("Lower leaf surface not captured")
            reqs.append("lowerLeaf")
        if not checklist.get("wholePlant"):
            issues.append("Whole-plant context missing")
            reqs.append("wholePlant")
        if not checklist.get("lightingOk"):
            issues.append("Poor lighting / glare")
            reqs.append("retake in better light")
        passed = (
            cov >= cq["minCoverage"]
            and checklist.get("leafClose", False)
            and sum(1 for v in cq["secondaryViews"] if checklist.get(v)) >= cq["minSecondaryViews"]
            and checklist.get("lightingOk", False)
        )
        return {"coverageScore": cov, "passed": passed, "issues": issues if not passed else [], "recaptureRequests": reqs if not passed else []}

    # ---------------- deterministic triage ----------------
    def diagnose(self, crop: str, symptom: str, checklist: dict[str, bool], at: str) -> dict[str, Any]:
        tax = self.taxonomy
        rows = tax["diagnosisTable"].get(f"{crop}:{symptom}")
        if rows is None:
            rows = [{"conditionId": "unknown", "base": 0.50}, {"conditionId": "healthy", "base": 0.30}, {"conditionId": "unknown", "base": 0.20}]
        t = self.policy["triage"]
        penalty = 0.0 if (checklist.get("lowerLeaf") and checklist.get("wholePlant")) else t["missingViewsScorePenalty"]
        cands: list[dict[str, Any]] = []
        for i, r in enumerate(rows):
            score = r["base"]
            if i == 0:
                score = round(max(0.05, score - penalty), 2)
            cid = r["conditionId"]
            cands.append({
                "conditionId": cid,
                "label": self.conditions[cid]["labelEn"],
                "simConfidence": score,
                "reasons": tax["reasons"].get(cid, tax["reasons"]["unknown"]),
                "missingEvidence": tax["missingEvidence"].get(cid, tax["missingEvidence"]["_default"]),
            })
        lead = cands[0]
        second = cands[1] if len(cands) > 1 else {"simConfidence": 0}
        margin = round(lead["simConfidence"] - second["simConfidence"], 2)
        high_spread = self.conditions[lead["conditionId"]]["spreadRisk"] == "high" and lead["simConfidence"] >= t["highSpreadRiskMinScore"]
        if lead["conditionId"] == "unknown" and lead["simConfidence"] >= t["abstainOtherThreshold"]:
            routing = {"decision": "abstain", "reason": "Out-of-distribution indicator: no candidate strong enough. Expert must decide; never forced into a known label."}
        elif lead["simConfidence"] >= t["autonomousMinScore"] and margin >= t["autonomousMinMargin"] and not high_spread:
            routing = {"decision": "autonomous", "reason": "Lead score and margin exceed prototype autonomous thresholds; safe advisory may issue without expert."}
        else:
            routing = {"decision": "expert", "reason": f"Lead score {lead['simConfidence']:.2f} below autonomous threshold {t['autonomousMinScore']} or margin {margin:.2f} too small; expert review required."}
        return {
            "provider": "demo-rules", "modelVersion": "0.1.0-demo", "provenance": "SIMULATED",
            "at": at, "crop": crop, "symptomCategory": symptom,
            "candidates": cands, "margin": margin, "routing": routing,
            "highSpreadRisk": high_spread, "escalationRequired": bool(high_spread and t["highSpreadRiskEscalates"]),
            "recommendedNext": tax["missingEvidence"].get(lead["conditionId"], tax["missingEvidence"]["_default"]) if routing["decision"] != "abstain" else tax["missingEvidence"]["unknown"],
            "thresholdsUsed": {"autonomousMinScore": t["autonomousMinScore"], "autonomousMinMargin": t["autonomousMinMargin"], "abstainOtherThreshold": t["abstainOtherThreshold"]},
            "note": "Simulated scores from deterministic demo rules — not measured model accuracy.",
        }

    # ---------------- outbreak scoring ----------------
    def outbreak_score(self, cluster: dict[str, Any], cases: list[dict[str, Any]]) -> dict[str, Any]:
        members = [c for c in cases if c["id"] in cluster["memberCaseIds"] and c["state"] != "CLOSED_DUPLICATE"]
        verified = [c for c in members if c.get("expertConfirmedCondition") == cluster["conditionId"]]
        member_count = max(1, len(members))
        verified_ratio = len(verified) / member_count
        ob = self.policy["outbreak"]
        w = ob["weights"]
        sig = cluster["seedSignals"]
        components = {
            "verifiedRatio": round(verified_ratio, 3),
            "spatialDensity": sig["spatialDensity"],
            "temporalGrowth": sig["temporalGrowth"],
            "cropStageCompat": sig["cropStageCompat"],
            "weatherSuitability": cluster["weatherSuitability"],
            "severityIndex": sig["severityIndex"],
        }
        positive = sum(w[k] * components[k] for k in w)
        penalty = ob["duplicatePenaltyWeight"] * sig["duplicatePenalty"]
        score = round(max(0.0, min(100.0, 100.0 * (positive - penalty))), 1)
        if cluster["status"] == "DISMISSED":
            status = "DISMISSED"
        elif score >= ob["thresholds"]["verifiedOutbreak"]:
            status = "VERIFIED"
        elif score >= ob["thresholds"]["suspected"]:
            status = "SUSPECTED"
        else:
            status = "WATCH"
        top = sorted(((k, w[k] * components[k]) for k in w), key=lambda x: -x[1])[:2]
        explanation = (
            f"Score {score} driven mainly by {top[0][0]} ({components[top[0][0]]}) and {top[1][0]} ({components[top[1][0]]}); "
            f"{len(verified)}/{member_count} member cases expert-verified; duplicate penalty {sig['duplicatePenalty']}."
        )
        return {
            "clusterId": cluster["id"], "score": score, "status": status,
            "components": components, "weights": w, "duplicatePenalty": sig["duplicatePenalty"],
            "verifiedCount": len(verified), "memberCount": member_count,
            "explanation": explanation, "provenance": "SIMULATED",
        }

    # ---------------- expert queue priority ----------------
    def expert_priority(self, case: dict[str, Any]) -> tuple[float, str]:
        """Deterministic queue priority. Higher = review sooner. Documented policy."""
        score = 10.0
        reasons: list[str] = []
        d = case.get("diagnosis")
        if d:
            if d.get("highSpreadRisk"):
                score += 50
                reasons.append("high-spread-risk candidate")
            if d.get("escalationRequired"):
                score += 25
                reasons.append("escalation required by policy")
            lead = d["candidates"][0]
            if lead["conditionId"] == "unknown":
                score += 20
                reasons.append("abstained / out-of-distribution")
            if lead["simConfidence"] < 0.5:
                score += 10
                reasons.append("very low lead score")
            reasons.append(f"lead {lead['conditionId']} (simulated {lead['simConfidence']:.2f})")
        if case.get("pendingSync"):
            score -= 5
            reasons.append("sync pending")
        return min(100.0, score), "; ".join(reasons)

    # ---------------- mission representative ordering ----------------
    def representative_order(self, cluster: dict[str, Any], cases: list[dict[str, Any]], limit: int = 3) -> list[str]:
        members = [c for c in cases if c["id"] in cluster["memberCaseIds"] and c["state"] != "CLOSED_DUPLICATE"]

        def dist(c: dict[str, Any]) -> float:
            return haversine_km(c["lat"], c["lon"], cluster["centerLat"], cluster["centerLon"])

        unverified = sorted([c for c in members if c.get("expertConfirmedCondition") != cluster["conditionId"]], key=dist)
        verified = sorted([c for c in members if c.get("expertConfirmedCondition") == cluster["conditionId"]], key=dist)
        return [c["id"] for c in (unverified + verified)][:limit]

File: apps/web/src/lib/engine.ts
/**
 * Deterministic demo engine (TypeScript) — mirrors apps/api/app/engine.py and
 * data/demo/generate_seed.py exactly. NOT a trained model: all scores are
 * simulated and labelled as such. Thresholds are prototype policies
 * (data/demo/policy.json), not validated agronomic thresholds.
 */
import type {
  CaptureChecklist, CaptureQuality, Case, DiagnosisResult, OutbreakCluster, OutbreakScoreBreakdown,
} from "@contracts";
import { CONDITIONS, POLICY, TAXONOMY } from "./seed";

export function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const r = 6371.0;
  const p1 = (lat1 * Math.PI) / 180;
  const p2 = (lat2 * Math.PI) / 180;
  const dp = ((lat2 - lat1) * Math.PI) / 180;
  const dl = ((lon2 - lon1) * Math.PI) / 180;
  const a = Math.sin(dp / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2;
  return 2 * r * Math.asin(Math.sqrt(a));
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

export function captureQuality(checklist: CaptureChecklist): CaptureQuality {
  const cq = POLICY.captureQuality;
  const w = cq.weights as Record<string, number>;
  const cov = round2(
    (checklist.leafClose ? w.leafClose : 0) +
    (checklist.lowerLeaf ? w.lowerLeaf : 0) +
    (checklist.wholePlant ? w.wholePlant : 0) +
    (checklist.lightingOk ? w.lightingOk : 0),
  );
  const issues: string[] = [];
  const reqs: string[] = [];
  if (!checklist.leafClose) { issues.push("Missing close-up of affected leaf"); reqs.push("leafClose"); }
  if (!checklist.lowerLeaf) { issues.push("Lower leaf surface not captured"); reqs.push("lowerLeaf"); }
  if (!checklist.wholePlant) { issues.push("Whole-plant context missing"); reqs.push("wholePlant"); }
  if (!checklist.lightingOk) { issues.push("Poor lighting / glare"); reqs.push("retake in better light"); }
  const secondary = [checklist.lowerLeaf, checklist.wholePlant].filter(Boolean).length;
  const passed = cov >= cq.minCoverage && checklist.leafClose && secondary >= cq.minSecondaryViews && checklist.lightingOk;
  return { coverageScore: cov, passed, issues: passed ? [] : issues, recaptureRequests: passed ? [] : reqs };
}

export function diagnose(crop: string, symptom: string, checklist: CaptureChecklist, at: string): DiagnosisResult {
  const key = `${crop}:${symptom}`;
  const table = TAXONOMY.diagnosisTable as unknown as Record<string, { conditionId: string; base: number }[]>;
  const rows = table[key] ?? [
    { conditionId: "unknown", base: 0.5 }, { conditionId: "healthy", base: 0.3 }, { conditionId: "unknown", base: 0.2 },
  ];
  const t = POLICY.triage;
  const penalty = checklist.lowerLeaf && checklist.wholePlant ? 0 : t.missingViewsScorePenalty;
  const reasons = TAXONOMY.reasons as Record<string, string[]>;
  const missing = TAXONOMY.missingEvidence as Record<string, string[]>;
  const candidates = rows.map((r, i) => {
    const score = i === 0 ? round2(Math.max(0.05, r.base - penalty)) : r.base;
    return {
      conditionId: r.conditionId,
      label: CONDITIONS[r.conditionId]?.labelEn ?? r.conditionId,
      simConfidence: score,
      reasons: reasons[r.conditionId] ?? reasons.unknown,
      missingEvidence: missing[r.conditionId] ?? missing._default,
    };
  });
  const lead = candidates[0];
  const second = candidates[1] ?? { simConfidence: 0 };
  const margin = round2(lead.simConfidence - second.simConfidence);
  const highSpread = (CONDITIONS[lead.conditionId]?.spreadRisk === "high") && lead.simConfidence >= t.highSpreadRiskMinScore;
  let routing: DiagnosisResult["routing"];
  if (lead.conditionId === "unknown" && lead.simConfidence >= t.abstainOtherThreshold) {
    routing = { decision: "abstain", reason: "Out-of-distribution indicator: no candidate strong enough. Expert must decide; never forced into a known label." };
  } else if (lead.simConfidence >= t.autonomousMinScore && margin >= t.autonomousMinMargin && !highSpread) {
    routing = { decision: "autonomous", reason: "Lead score and margin exceed prototype autonomous thresholds; safe advisory may issue without expert." };
  } else {
    routing = { decision: "expert", reason: `Lead score ${lead.simConfidence.toFixed(2)} below autonomous threshold ${t.autonomousMinScore} or margin ${margin.toFixed(2)} too small; expert review required.` };
  }
  return {
    provider: "demo-rules", modelVersion: "0.1.0-demo", provenance: "SIMULATED",
    at, crop, symptomCategory: symptom, candidates, margin, routing,
    highSpreadRisk: highSpread, escalationRequired: highSpread && t.highSpreadRiskEscalates,
    recommendedNext: routing.decision === "abstain" ? missing.unknown : (missing[lead.conditionId] ?? missing._default),
    thresholdsUsed: { autonomousMinScore: t.autonomousMinScore, autonomousMinMargin: t.autonomousMinMargin, abstainOtherThreshold: t.abstainOtherThreshold },
    note: "Simulated scores from deterministic demo rules — not measured model accuracy.",
  };
}

export function outbreakScore(cluster: OutbreakCluster, cases: Case[]): OutbreakScoreBreakdown {
  const members = cases.filter((c) => cluster.memberCaseIds.includes(c.id) && c.state !== "CLOSED_DUPLICATE");
  const verified = members.filter((c) => c.expertConfirmedCondition === cluster.conditionId);
  const memberCount = Math.max(1, members.length);
  const verifiedRatio = verified.length / memberCount;
  const ob = POLICY.outbreak;
  const w = ob.weights;
  const sig = cluster.seedSignals;
  const components = {
    verifiedRatio: Math.round(verifiedRatio * 1000) / 1000,
    spatialDensity: sig.spatialDensity,
    temporalGrowth: sig.temporalGrowth,
    cropStageCompat: sig.cropStageCompat,
    weatherSuitability: cluster.weatherSuitability,
    severityIndex: sig.severityIndex,
  };
  const positive = Object.entries(w).reduce((acc, [k, wt]) => acc + wt * components[k as keyof typeof components], 0);
  const penalty = ob.duplicatePenaltyWeight * sig.duplicatePenalty;
  const score = Math.round(Math.max(0, Math.min(100, 100 * (positive - penalty))) * 10) / 10;
  const status = cluster.status === "DISMISSED" ? "DISMISSED"
    : score >= ob.thresholds.verifiedOutbreak ? "VERIFIED"
    : score >= ob.thresholds.suspected ? "SUSPECTED" : "WATCH";
  const top = Object.entries(w)
    .map(([k, wt]) => [k, wt * components[k as keyof typeof components]] as const)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 2);
  const explanation = `Score ${score} driven mainly by ${top[0][0]} (${components[top[0][0] as keyof typeof components]}) and ${top[1][0]} (${components[top[1][0] as keyof typeof components]}); ${verified.length}/${memberCount} member cases expert-verified; duplicate penalty ${sig.duplicatePenalty}.`;
  return {
    clusterId: cluster.id, score, status, components, weights: w,
    duplicatePenalty: sig.duplicatePenalty, verifiedCount: verified.length,
    memberCount: memberCount, explanation, provenance: "SIMULATED",
  };
}

export function expertPriority(c: Case): { score: number; reason: string } {
  let score = 10;
  const reasons: string[] = [];
  const d = c.diagnosis;
  if (d) {
    if (d.highSpreadRisk) { score += 50; reasons.push("high-spread-risk candidate"); }
    if (d.escalationRequired) { score += 25; reasons.push("escalation required by policy"); }
    const lead = d.candidates[0];
    if (lead.conditionId === "unknown") { score += 20; reasons.push("abstained / out-of-distribution"); }
    if (lead.simConfidence < 0.5) { score += 10; reasons.push("very low lead score"); }
    reasons.push(`lead ${lead.conditionId} (simulated ${lead.simConfidence.toFixed(2)})`);
  }
  if (c.pendingSync) { score -= 5; reasons.push("sync pending"); }
  return { score: Math.min(100, score), reason: reasons.join("; ") };
}

export function representativeOrder(cluster: OutbreakCluster, cases: Case[], limit = 3): string[] {
  const members = cases.filter((c) => cluster.memberCaseIds.includes(c.id) && c.state !== "CLOSED_DUPLICATE");
  const dist = (c: Case) => haversineKm(c.lat, c.lon, cluster.centerLat, cluster.centerLon);
  const unverified = members.filter((c) => c.expertConfirmedCondition !== cluster.conditionId).sort((a, b) => dist(a) - dist(b));
  const verified = members.filter((c) => c.expertConfirmedCondition === cluster.conditionId).sort((a, b) => dist(a) - dist(b));
  return [...unverified, ...verified].slice(0, limit).map((c) => c.id);
}
